pyomo/contrib/piecewise/transform/disaggregated_logarithmic.py

```diff
@@ -0,0 +1,199 @@
+
+        # The lambda variables \lambda_{P,v} are indexed by the simplex and the point in it
+        transBlock.lambdas = Var(
+            transBlock.simplex_indices, transBlock.simplex_point_indices, bounds=(0, 1)
+        )
+
+        # Numbered citations are from Vielma et al 2010, Mixed-Integer Models
+        # for Nonseparable Piecewise-Linear Optimization
+
+        # Sum of all lambdas is one (6b)
+        transBlock.convex_combo = Constraint(
+            expr=sum(
+                transBlock.lambdas[P, v]
+                for P in transBlock.simplex_indices
+                for v in transBlock.simplex_point_indices
+            )
+            == 1
+        )
+
+        # The branching rules, establishing using the binaries that only one simplex's lambda
+        # coefficients may be nonzero
+        # Enabling lambdas when binaries are on
+        @transBlock.Constraint(transBlock.log_simplex_indices)  # (6c.1)
+        def simplex_choice_1(b, l):
+            return (
+                sum(
+                    transBlock.lambdas[P, v]
+                    for P in transBlock.P_plus[l]
+                    for v in transBlock.simplex_point_indices
+                )
+                <= transBlock.binaries[l]
+            )
+
+        # Disabling lambdas when binaries are on
+        @transBlock.Constraint(transBlock.log_simplex_indices)  # (6c.2)
+        def simplex_choice_2(b, l):
+            return (
+                sum(
+                    transBlock.lambdas[P, v]
+                    for P in transBlock.P_0[l]
+                    for v in transBlock.simplex_point_indices
+                )
+                <= 1 - transBlock.binaries[l]
+            )
+
+        # for i, (simplex, pwlf) in enumerate(choices):
+        # x_i = sum(lambda_P,v v_i, P in polytopes, v in V(P))
+        @transBlock.Constraint(transBlock.dimension_indices)  # (6a.1)
+        def x_constraint(b, i):
+            return pw_expr.args[i] == sum(
+                transBlock.lambdas[P, v]
+                * pw_linear_func._points[idx_to_simplex[P][v]][i]
+                for P in transBlock.simplex_indices
+                for v in transBlock.simplex_point_indices
+            )
+
+        # Make the substitute Var equal the PWLE (6a.2)
+        transBlock.set_substitute = Constraint(
+            expr=substitute_var
+            == sum(
+                transBlock.lambdas[P, v]
+                * linear_func(*pw_linear_func._points[idx_to_simplex[P][v]])
+                for v in transBlock.simplex_point_indices
+                for (P, linear_func) in simplex_indices_and_lin_funcs
+            )
+        )
+
+        return substitute_var
+
+    # Not a Gray code, just a regular binary representation
+    # TODO test the Gray codes too
+    def _get_binary_vector(self, num, length):
+        if num != 0 and ceil(log2(num)) > length:
+            raise DeveloperError("Invalid input in _get_binary_vector")
```

Comment:
Do you need this check? Since it's a private method, giving valid input is in your control.

Reply:
You're right, it isn't necessary. I removed that check and added a comment instead.